File: bindings/soth-node/examples/batch_proof.js
// Fire 20 cheap Anthropic calls concurrently through the Node SDK
// and watch the shipper's POST lines so we can count how many HTTP
// requests actually leave the box.
//
// Mirrors examples/batch_proof.py from soth-py — both bindings sit
// on the same Rust shipper in soth-sdk-core, so the wire format and
// batching cadence should be identical.

const Anthropic = require('@anthropic-ai/sdk');
const soth = require('..');

if (!process.env.SOTH_ORG_ID || !process.env.SOTH_API_KEY) {
  console.error('SOTH_ORG_ID and SOTH_API_KEY are required (find them in ~/.soth/soth.yaml)');
  process.exit(1);
}
const ORG_ID = process.env.SOTH_ORG_ID;
const SOTH_API_KEY = process.env.SOTH_API_KEY;
const TELEMETRY_ENDPOINT = process.env.SOTH_TELEMETRY_ENDPOINT
  || 'https://ingest.soth.ai/v1/edge/telemetry/batch';
const MODEL = 'claude-haiku-4-5-20251001';
const N_CALLS = 20;

async function main() {
  if (!process.env.ANTHROPIC_API_KEY) {
    console.error('ANTHROPIC_API_KEY not set');
    process.exit(1);
  }

  soth.init({
    apiKey: SOTH_API_KEY,
    orgId: ORG_ID,
    telemetryEndpoint: TELEMETRY_ENDPOINT,
  });
  const state = soth.instrument({ providers: ['anthropic'] });
  console.log('instrumentation:', state);

  const client = new Anthropic.default();

  const fire = async (i) => {
    const msg = await client.messages.create({
      model: MODEL,
      max_tokens: 8,
      messages: [{ role: 'user', content: `Reply with the number ${i}, nothing else.` }],
    });
    return msg.usage.output_tokens;
  };

  const t0 = Date.now();
  const outs = await Promise.all(Array.from({ length: N_CALLS }, (_, i) => fire(i)));
  const dt = (Date.now() - t0) / 1000;
  console.log(`\n>>> fired ${N_CALLS} calls in ${dt.toFixed(2)}s, total output tokens: ${outs.reduce((a, b) => a + b, 0)}`);
  console.log('>>> sleeping 12s to let shipper drain (BATCH_WINDOW=5s)…');
  await new Promise((r) => setTimeout(r, 12_000));
  soth.shutdown();
  console.log('>>> shutdown complete (forces final-drain POST)');
}

main().catch((e) => { console.error(e); process.exit(1); });

File: bindings/soth-py/examples/batch_proof.py
"""Fire 20 cheap Anthropic calls concurrently through the SDK and
print the shipper's POST lines so we can count how many HTTP requests
actually leave the box.

If batching works, 20 events should land in 1–2 POSTs (one if they
all queue inside a single 5s batch window, two if some land after
the first window fires).
"""

from __future__ import annotations

import os
import sys
import time
from concurrent.futures import ThreadPoolExecutor

import anthropic
import soth

ORG_ID = os.environ["SOTH_ORG_ID"]
SOTH_API_KEY = os.environ["SOTH_API_KEY"]
TELEMETRY_ENDPOINT = os.environ.get(
    "SOTH_TELEMETRY_ENDPOINT", "https://ingest.soth.ai/v1/edge/telemetry/batch"
)
MODEL = "claude-haiku-4-5-20251001"
N_CALLS = 20


def fire(client: anthropic.Anthropic, i: int) -> int:
    msg = client.messages.create(
        model=MODEL,
        max_tokens=8,
        messages=[{"role": "user", "content": f"Reply with the number {i}, nothing else."}],
    )
    return msg.usage.output_tokens


def main() -> int:
    if not os.environ.get("ANTHROPIC_API_KEY"):
        print("ANTHROPIC_API_KEY not set", file=sys.stderr)
        return 1

    soth.init(api_key=SOTH_API_KEY, org_id=ORG_ID, telemetry_endpoint=TELEMETRY_ENDPOINT)
    soth.instrument(providers=["anthropic"])
    client = anthropic.Anthropic()

    t0 = time.time()
    with ThreadPoolExecutor(max_workers=10) as ex:
        outs = list(ex.map(lambda i: fire(client, i), range(N_CALLS)))
    dt = time.time() - t0
    print(f"\n>>> fired {N_CALLS} calls in {dt:.2f}s, total output tokens: {sum(outs)}")
    print(">>> sleeping 12s to let shipper drain (BATCH_WINDOW=5s)…")
    time.sleep(12)
    soth.shutdown()
    print(">>> shutdown complete (forces final-drain POST)")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
